File: python/src/acp/negotiate.py
from __future__ import annotations

from .types import Capability, ProtocolRange, parse_version

ENCODING_PREFERENCE = ("cbor", "json")
HEARTBEAT_MIN_MS = 100
HEARTBEAT_MAX_MS = 60_000
MESSAGE_BYTES_MIN = 256
MESSAGE_BYTES_MAX = 1_048_576


class VersionError(ValueError):
    """Handshake version negotiation failed."""


def select_encoding(client: list[str], server: list[str]) -> str:
    client_set = set(client)
    server_set = set(server)
    for enc in ENCODING_PREFERENCE:
        if enc in client_set and enc in server_set:
            return enc
    raise VersionError("no common encoding")


def intersect_capabilities(local: list[Capability], peer: list[Capability]) -> list[Capability]:
    peer_map = {c.id: c.version for c in peer}
    out: list[Capability] = []
    for cap in local:
        if cap.id not in peer_map:
            continue
        other = peer_map[cap.id]
        try:
            lv, ov = parse_version(cap.version), parse_version(other)
        except ValueError:
            continue
        if lv[0] != ov[0]:
            continue
        chosen = cap.version if lv <= ov else other
        out.append(Capability(cap.id, chosen))
    return out


def validate_heartbeat(ms: int) -> int:
    if not HEARTBEAT_MIN_MS <= ms <= HEARTBEAT_MAX_MS:
        raise VersionError(f"heartbeat_interval_ms out of bounds: {ms}")
    return ms


def validate_max_message_bytes(n: int) -> int:
    if not MESSAGE_BYTES_MIN <= n <= MESSAGE_BYTES_MAX:
        raise VersionError(f"max_message_bytes out of bounds: {n}")
    return n


def version_leq(actual: str, maximum: str) -> bool:
    return parse_version(actual) <= parse_version(maximum)


def select_version(client: ProtocolRange, server: ProtocolRange) -> str:
    """Highest shared minor within the same major, or raise VersionError."""
    cmin, cmax = parse_version(client.min), parse_version(client.max)
    smin, smax = parse_version(server.min), parse_version(server.max)
    if cmin[0] != smin[0] or cmin[0] != cmax[0] or smin[0] != smax[0]:
        raise VersionError("protocol major mismatch or malformed range")
    lo = max(cmin[1], smin[1])
    hi = min(cmax[1], smax[1])
    if lo > hi:
        raise VersionError("empty protocol intersection")
    return f"{cmin[0]}.{hi}"


def version_at_least(actual: str, required: str) -> bool:
    a = parse_version(actual)
    r = parse_version(required)
    return a >= r

File: rust/acp-session/src/negotiate.rs
use acp_model::{Capability, ProtocolRange};

pub const ENCODING_PREFERENCE: [&str; 2] = ["cbor", "json"];
pub const HEARTBEAT_MIN_MS: i64 = 100;
pub const HEARTBEAT_MAX_MS: i64 = 60_000;

#[derive(Debug)]
pub struct NegotiateError(pub String);

impl std::fmt::Display for NegotiateError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for NegotiateError {}

fn parse_version(text: &str) -> Result<(u32, u32), NegotiateError> {
    let mut parts = text.split('.');
    let major = parts
        .next()
        .and_then(|s| s.parse().ok())
        .ok_or_else(|| NegotiateError("bad version".into()))?;
    let minor = parts
        .next()
        .and_then(|s| s.parse().ok())
        .ok_or_else(|| NegotiateError("bad version".into()))?;
    if parts.next().is_some() {
        return Err(NegotiateError("bad version".into()));
    }
    Ok((major, minor))
}

pub fn select_version(
    client: &ProtocolRange,
    server: &ProtocolRange,
) -> Result<String, NegotiateError> {
    let cmin = parse_version(&client.min)?;
    let cmax = parse_version(&client.max)?;
    let smin = parse_version(&server.min)?;
    let smax = parse_version(&server.max)?;
    if cmin.0 != smin.0 || cmin.0 != cmax.0 || smin.0 != smax.0 || cmin > cmax || smin > smax {
        return Err(NegotiateError(
            "protocol major mismatch or malformed range".into(),
        ));
    }
    let lo = cmin.1.max(smin.1);
    let hi = cmax.1.min(smax.1);
    if lo > hi {
        return Err(NegotiateError("empty protocol intersection".into()));
    }
    Ok(format!("{}.{}", cmin.0, hi))
}

pub fn select_encoding(client: &[String], server: &[String]) -> Result<String, NegotiateError> {
    for enc in ENCODING_PREFERENCE {
        if client.iter().any(|e| e == enc) && server.iter().any(|e| e == enc) {
            return Ok(enc.to_string());
        }
    }
    Err(NegotiateError("no common encoding".into()))
}

pub fn intersect_profiles(local: &[String], peer: &[String]) -> Vec<String> {
    local
        .iter()
        .filter(|name| peer.iter().any(|p| p == *name))
        .cloned()
        .collect()
}

pub fn intersect_capabilities(local: &[Capability], peer: &[Capability]) -> Vec<Capability> {
    let mut out = Vec::new();
    for cap in local {
        if let Some(other) = peer.iter().find(|p| p.id == cap.id) {
            let Ok(lv) = parse_version(&cap.version) else {
                continue;
            };
            let Ok(ov) = parse_version(&other.version) else {
                continue;
            };
            if lv.0 != ov.0 {
                continue;
            }
            let version = if lv <= ov {
                cap.version.clone()
            } else {
                other.version.clone()
            };
            out.push(Capability {
                id: cap.id.clone(),
                version,
            });
        }
    }
    out
}

pub fn version_at_least(actual: &str, required: &str) -> bool {
    match (parse_version(actual), parse_version(required)) {
        (Ok(a), Ok(r)) => a >= r,
        _ => false,
    }
}

pub fn version_leq(actual: &str, maximum: &str) -> bool {
    match (parse_version(actual), parse_version(maximum)) {
        (Ok(a), Ok(m)) => a <= m,
        _ => false,
    }
}

pub fn validate_heartbeat(ms: i64) -> Result<i64, NegotiateError> {
    if (HEARTBEAT_MIN_MS..=HEARTBEAT_MAX_MS).contains(&ms) {
        Ok(ms)
    } else {
        Err(NegotiateError("heartbeat out of bounds".into()))
    }
}
